kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/WorkflowViewStub.kt

```diff
@@ -0,0 +1,62 @@
+package com.squareup.workflow.ui
+
+import android.content.Context
+import android.util.AttributeSet
+import android.view.View
+import android.view.ViewGroup
+
+/**
+ * A placeholder [View] that can replace itself with ones driven by workflow renderings,
+ * similar to [android.view.ViewStub].
+ */
+class WorkflowViewStub(
+  context: Context,
+  attributeSet: AttributeSet? = null
+) : View(context, attributeSet) {
+  init {
+    setWillNotDraw(true)
+  }
+
+  /**
+   * By starting as our own delegate, we make the parent checking code in [update] simpler.
+   */
+  private var _actual: View = this
+
+  /**
+   * On-demand access to the delegate established by the last call to [update],
+   * or null if none has been set yet.
+   */
+  val actual: View? get() = _actual.takeIf { it !is WorkflowViewStub }
+
+  /**
+   * Replaces this view with one that can display [rendering]. If it
+   * has already been replaced, updates that replacement if it [canShowRendering].
+   * If it can't, a new replacement is put in place.
+   *
+   * @return the view that showed [rendering]
+   * @throws RuntimeException if [rendering] cannot be displayed, typically indicating
+   * a misconfigured [registry]
+   */
+  fun update(
+    rendering: Any,
+    registry: ViewRegistry
+  ): View {
+    _actual.takeIf { it.canShowRendering(rendering) }
+        ?.let {
+          it.showRendering(rendering)
+          return it
+        }
+
+    return when (val parent = _actual.parent) {
+      is ViewGroup -> registry.buildView(rendering, parent).also { newView ->
+        val index = parent.indexOfChild(_actual)
+        parent.removeView(_actual)
+
+        _actual.layoutParams
+            ?.let { parent.addView(newView, index, it) }
+            ?: run { parent.addView(newView, index) }
+      }
```

Comment:
Nit: Maybe factor the body of this case out into a separate function `buildAndReplaceView` so this `when` is nice and scannable, especially since the `.also` at the end kind of throws you for a loop because you have to go back up to the top of the `when` to figure out what it actually refers to.